Allow specifing database backup paths.

libklein opened the pull request:
Resolves #3279.

Allow to specify a custom backup location & filename. The filename may contain patterns "{DB_NAME}" and "{TIME:xxx}" which will be replaced with the database name and the current time, formatted according to "xxx", respectively.

The backup path can be relative or absolute. The former creates the backup relative to the database file.

The following changes have been made:
- `backupDatabase()` now takes a destination file path as argument
- Propagated method signature changes to `saveAs`, `save` and `performSave` functions
- Added a `getDefault(key)` method to `Config`. This returns the default value.
- Added new config keys `backupFileNamePattern` and `backupFilePath`
- Added textfields to `ApplicationSettingsWidget` accordingly

## Limitations/open issues (please comment)
- Old databases are not removed when the backup location changes. I believe this is out of scope and should not be the responsibility of keepassxc. Especially since there is no way to handle relative paths properly if this is required.
- A section on the structure of `backupFileNamePattern` needs to created. Or do you think it suffices to explain it in a tooltip?

## Screenshots
![image](https://user-images.githubusercontent.com/42714034/136691539-2abe9d2b-df77-49d5-8ea9-4db1bb477207.png)

## Testing strategy
Modified database backup unit test accordingly.

## Type of change
- ✅ New feature (change that adds functionality)

libklein commented:
@ijordse  The current behavior is to name backups according to the pattern specified and save them under the given path. So a pattern of "{DB_NAME}-{TIME:dd.MM.yyyy}.backup" with a path of "/tmp" and db name "test" would save backups as "/tmp/test-11.10.2021.backup". Is that not what your comment suggests? Could you elaborate?

droidmonkey commented on src/core/Config.cpp:
```diff
@@ -61,6 +61,8 @@ static const QHash<Config::ConfigKey, ConfigDirective> configStrings = {
     {Config::AutoSaveNonDataChanges,{QS("AutoSaveNonDataChanges"), Roaming, true}},
     {Config::BackupBeforeSave,{QS("BackupBeforeSave"), Roaming, false}},
+    {Config::BackupFilePath,{QS("BackupFilePath"), Roaming, QString(".")}},
+    {Config::BackupFileNamePattern,{QS("BackupFilePathPattern"), Roaming, QString("{DB_NAME}-{TIME}.old")}},
```
To maintain consistency with older versions, this should default to {DB_NAME}.old.kdbx

libklein replied:
Done

droidmonkey commented on src/core/Database.cpp:
```diff
@@ -485,23 +486,26 @@ void Database::releaseData()
-    auto match = re.match(filePath);
-    auto backupFilePath = filePath;
+    // Ensure that the path to write to actually exists
+    auto parentDirectory = QDir(destinationFilePath.section(QDir::separator(), 0, -2));
```
Use QFileInfo::absolutePath instead of this.

libklein replied:
Fixed. Using QFileInfo whenever i parse the parent directory.

droidmonkey commented on src/gui/ApplicationSettingsWidget.cpp:
```diff
@@ -188,6 +214,13 @@ void ApplicationSettingsWidget::loadSettings()
+    if (backupFilePath != QString(".")) { // Hide ugly "." and use replacement text instead.
+        m_generalUi->backupFileDirectory->setText(backupFilePath);
+    }
+    m_generalUi->backupFileName->setText(config()->get(Config::BackupFileNamePattern).toString());
```
Combine these two fields into one. You should set the path and the filename pattern in the same text edit field. If there is only a path provided then use the default filename pattern. Lessens the settings clutter.

libklein replied:
I've combined the fields. This actually makes the UI much cleaner as well!

droidmonkey commented on src/gui/DatabaseWidget.cpp:
```diff
@@ -1881,11 +1881,45 @@ bool DatabaseWidget::performSave(QString& errorMessage, const QString& fileName)
+        backupFilePath = config()->get(Config::BackupFilePathPattern).toString();
+        // Fall back to default
+        if (backupFilePath.isEmpty()) {
+            config()->getDefault(Config::BackupFilePathPattern).toString();
```
Missing assignment to `backupFilePath`

droidmonkey commented on src/gui/DatabaseWidget.cpp:
```diff
@@ -1881,11 +1881,45 @@ bool DatabaseWidget::performSave(QString& errorMessage, const QString& fileName)
+            config()->getDefault(Config::BackupFilePathPattern).toString();
+        }
+        // Replace backup pattern
+        backupFilePath.replace(QString("{DB_NAME}"), m_db->metadata()->name());
```
This leaks database information. The placeholder should be `{DB_FILENAME}` which would be everything up to the extension (e.g., my.database_rocks.kdbx would resolve to my.database_rocks)

droidmonkey commented on src/gui/DatabaseWidget.cpp:
```diff
@@ -1881,11 +1881,45 @@ bool DatabaseWidget::performSave(QString& errorMessage, const QString& fileName)
+            auto dbDir = QFileInfo(m_db->filePath()).absolutePath();
+            backupFilePath = QDir::cleanPath(dbDir + QDir::separator() + backupFilePath);
+        }
+    }
```
All of this should be in Tools.cpp that way the CLI and other parts of the program (KeeShare) can participate in this feature as well in the future.

droidmonkey commented on src/cli/Add.cpp:
```diff
@@ -121,7 +121,7 @@ int Add::executeWithDatabase(QSharedPointer<Database> database, QSharedPointer<Q
 
     QString errorMessage;
-    if (!database->save(Database::Atomic, false, &errorMessage)) {
+    if (!database->save(Database::Atomic, QString(), &errorMessage)) {
```
For all of these replacements, use `{}` instead.

droidmonkey commented on tests/TestDatabase.cpp:
```diff
@@ -75,29 +75,26 @@ void TestDatabase::testSave()
 
     // Test save backups
+    TemporaryFile backupFile;
+    auto backupFilePath = backupFile.fileName();
     db->metadata()->setName("test4");
-    QVERIFY2(db->save(Database::Atomic, true, &error), error.toLatin1());
+    QVERIFY2(db->save(Database::Atomic, backupFilePath, &error), error.toLatin1());
     QVERIFY(!db->isModified());
```
Need to add tests to TestGui.cpp as well while modulating the backup file configuration setting.

droidmonkey commented on src/gui/ApplicationSettingsWidget.cpp:
```diff
@@ -28,11 +30,24 @@
+        }
+    }
+} // namespace
+
```
I don't like this here and seems unnecessary. If the user leaves a setting blank it should be handled appropriately elsewhere, but the setting is blank.

droidmonkey commented on src/gui/ApplicationSettingsWidget.cpp:
```diff
@@ -504,3 +534,16 @@ void ApplicationSettingsWidget::checkUpdatesToggled(bool checked)
+        FileDialog::instance()->getExistingDirectory(this, tr("Select backup storage directory"), QDir::homePath());
+    if (!backupDirectory.isEmpty()) {
+        m_generalUi->backupFilePath->setText(backupDirectory);
+    }
```
You should append the default backup filename to the selected directory.

libklein replied:
Should we not rather append the default backup file **path**? If the default is ever going to be a path rather than a filename that probably has a good reason.

droidmonkey replied:
Whatever the default value is should be appended. At this point that is a file name which is appropriate.

droidmonkey commented on src/gui/ApplicationSettingsWidget.cpp:
```diff
@@ -504,3 +534,16 @@ void ApplicationSettingsWidget::checkUpdatesToggled(bool checked)
+}
+void ApplicationSettingsWidget::showBackupFilePathHelp()
+{
+    QDesktopServices::openUrl(QUrl("https://keepassxc.org/docs/KeePassXC_UserGuide.html#_backing_up_a_database_file"));
```
Remove the help button, use a tooltip on the text edit instead. When you hover over it should mention the {DB_FILENAME} and {TIME...} parameters.

droidmonkey commented:
Minor cleanup. I removed the ConfigContext since it doesn't conform to our code standards and it really should be handled in the init() instead. Although I am getting a failure in the backup test locally but in autoreload test on ci.

libklein commented on tests/gui/TestGui.cpp:
```diff
@@ -108,9 +97,23 @@ void TestGui::initTestCase()
+    // Disable autosave so we can test the modified file indicator
+    config()->set(Config::AutoSaveAfterEveryChange, false);
+    config()->set(Config::AutoSaveOnExit, false);
+    config()->set(Config::AutoReloadOnChange, true);
```
Adding config()->set(Config::AutoReloadOnChange, true) in init() causes the autoreloadDatabase test to fail.

libklein commented:
> Minor cleanup. I removed the ConfigContext since it doesn't conform to our code standards and it really should be handled in the init() instead. Although I am getting a failure in the backup test locally but in autoreload test on ci.

Im not sure if this is a good idea, especially against the background of the test failure the last commit caused. I think we need a more robust solution here, it's way to easy to forget adding/removing respective config()->set calls when editing tests. 

I think it makes sense that the init() method is responsible for providing a clean test environment and would suggest simply reloading the config in init() or resetting it to defaults. In line with this, tests should not depend on the value of specific settings even after calls to init() but rather set those they depend on in the test body (thats essentially why the AutoreloadDatabase test failed).

libklein committed:
Allow specifing database backup paths.

Update database saveAs() and save() usages.

- Default backupFilePath is '{DB_FILENAME}.old.kdbx' to conform to existing standards
- Implement backupPathPattern tests.
- Show tooltip on how to format database backup location text field.

Code cleanup

* Remove ConfigContext addition

Remove dependency on default config setting in AutoreloadDatabase, Lines 361-368.

Create substituteBackupFilePathPattern in Tools.cpp.

Add tests for substituteBackupFilePathPattern.

libklein commented:
Would you mind labeling the PR with "hacktoberfest-accepted"? I'm fine with doing further modifications, but it's almost the end of October and i believe that it satisfies their expectations w.r.t. quality at this point.

I believe that the windows gui test failure is transient and not caused by the PR.

droidmonkey committed:
Code cleanup

droidmonkey commented:
I was actually working on this as you committed. I simplified the tools function a lot.

libklein commented on src/core/Tools.cpp:
```diff
@@ -379,12 +380,12 @@ namespace Tools
     }
 
-    QString
-    substituteBackupFilePathPattern(QString pattern, const QString& databasePath, QDateTime date, int maxSubstitutions)
```
Should we not maybe keep date and maxSubstitution parameters? Date generalizes the function which could be useful if we later want to use it in a different context. maxSubstitution is actually required for correctness if we add another regex replacement. As soon as we do that we could have loop by substituting A with B and vice versa (see the tests for a sketch). We could just hardcode the value for now though, or at least put a warning comment there.

droidmonkey replied:
I made it impossible to recurse regex by not doing a global match all at once. It matches the string after each replacement looking for the patterns from left to right, which will overwrite recursive ones with the substitution instead. 

It doesn't make any sense to have any other time besides now. A time in the past or future would either be a calculation from now (now - 1 day) or a an absolute date which wouldn't be a substitution.

libklein replied:
Sorry, i don't get how match() vs globalMatch() fixes a recursion error. We were replacing matches one-at-a-time (left-to-right) with globalMatch() in the previous code version, reevaluating after every substitution. Could you perhaps explain? Not saying that using match() is not more appropriate here, just curious what i am missing :sweat_smile:.

I was trying to address the following problem with the maxSubstitutions parameter: assume that we modify the function to add other patterns/use it in a different context (it's in the Tool namespace now, so i guess that will probably happen at some point). Maybe we add a placeholder that substitutes some DB Attribute, like the Database description or name. Then, setting the description to {DB_ATTR:DESC} and supplying regex "{DB_ATTR:DESC}" would cause an infinite loop because "{DB_ATTR:DESC}" is replaced with itself.

Again, this cannot happen right now. If we still want to address this issue and make the function more robust for future modifications we could either i) add the maxSubstitutions parameter, or ii) advance the starting point of match() in the new string based on the current index and length of the consumed substring + replacement to not consider already replaced placeholders.

droidmonkey replied:
Globalmatch finds all matches at once, including recursive ones, and then you iterate that match list (QRegularExpressionMatchIterator). This is bad when you are replacing text at each occurrence because you modify the string and screw up the positioning. You were finding them one at a time by calling global match every loop, but that unnecessarily complicates the code.

I am not worried about some future modification of this function. If it is modified to maybe introduce replacing something that can cause recursion, then it can be beefed up at that time.

https://doc.qt.io/qt-5/qregularexpression.html#global-matching

libklein replied:
> I made it impossible to recurse regex by not doing a global match all at once. It matches the string after each replacement looking for the patterns from left to right, which will overwrite recursive ones with the substitution instead.

I get that global match finds all matches at once and that it's unnessesary in this case and we should use match() instead. But even when using globalMatch() it is impossible to recurse regex, no? I don't see how substituting match() for globalMatch() modifies the behavior of the function given that i was only using the first match found previously.

droidmonkey replied:
This is true based on your previous implementation since you re-performed the global match after each replacement.

droidmonkey commented:
I'm going to look into the test failure...

libklein commented on tests/gui/TestGui.cpp:
```diff
@@ -1778,22 +1768,18 @@ void TestGui::addCannedEntries()
-    }
-
-    checkDatabase(dbFileName, m_db->metadata()->name());
+    checkDatabase(fileName.isEmpty() ? m_dbFileName : fileName, m_db->metadata()->name());
```
Should be m_dbFilePath.

s-KaiNet commented:
Thank you for this feature! That was the only missing thing that prevented me to migrate from classic KeePass.
